src/cart/crt_init.c

```diff
@@ -468,37 +486,24 @@ crt_initialized()
 void
 crt_plugin_fini(void)
 {
-	struct crt_prog_cb_priv		*prog_cb_priv;
-	struct crt_timeout_cb_priv	*timeout_cb_priv;
-	struct crt_event_cb_priv	*event_cb_priv;
-	int				 i;
+	int i;
 
 	D_ASSERT(crt_plugin_gdata.cpg_inited == 1);
 
+	crt_plugin_gdata.cpg_inited = 0;
+
 	for (i = 0; i < CRT_SRV_CONTEXT_NUM; i++) {
-		while ((prog_cb_priv = d_list_pop_entry(
-					&crt_plugin_gdata.cpg_prog_cbs[i],
-					struct crt_prog_cb_priv,
-					cpcp_link))) {
-			D_FREE(prog_cb_priv);
-		}
+		D_FREE(crt_plugin_gdata.cpg_prog_cbs[i]);
+		D_FREE(crt_plugin_gdata.cpg_prog_cbs_old[i]);
 	}
 
-	while ((timeout_cb_priv = d_list_pop_entry(&crt_plugin_gdata.cpg_timeout_cbs,
-						   struct crt_timeout_cb_priv,
-						   ctcp_link))) {
-		D_FREE(timeout_cb_priv);
-	}
-	while ((event_cb_priv = d_list_pop_entry(&crt_plugin_gdata.cpg_event_cbs,
-						 struct crt_event_cb_priv,
-						 cecp_link))) {
-		D_FREE(event_cb_priv);
-	}
+	D_FREE(crt_plugin_gdata.cpg_timeout_cbs);
+	D_FREE(crt_plugin_gdata.cpg_timeout_cbs_old);
 
-	for (i = 0; i < CRT_SRV_CONTEXT_NUM; i++)
-		D_RWLOCK_DESTROY(&crt_plugin_gdata.cpg_prog_rwlock[i]);
-	D_RWLOCK_DESTROY(&crt_plugin_gdata.cpg_timeout_rwlock);
-	D_RWLOCK_DESTROY(&crt_plugin_gdata.cpg_event_rwlock);
+	D_FREE(crt_plugin_gdata.cpg_event_cbs);
+	D_FREE(crt_plugin_gdata.cpg_event_cbs_old);
+
+	D_MUTEX_DESTROY(&crt_plugin_gdata.cpg_mutex);
```

Comment:
Could you please elaborate on what you are replace all those RWLOCK with just a single lock? Sounds like we are moving in the wrong direction.

Reply:
This lock is used only for add/remove a callback. This happens very rare. Therefore I think it not make sence to use separate locks for each type of callback. Do you still insist to use separate locks?